### ISSUE TITLE
Discussion: advantage to using sync over async for I/O?

### ISSUE THREAD

**Issue by roji** [dotnet/corefx#5044]
Continuation of discussion from #4868 

>>Without AuthenticateAsClient(), Open() must resort to calling an async method internally, which means that it's no longer purely synchronous...
>The code above is as synchronous as the old implementation. When you call synchronous I/O on something like NetworkStream, internally, the thread will be waiting on an event for the operation to finish. Even more broadly: in reality there is no such thing as synchronous I/O in Windows.

Thanks for this detail. I was aware that at least in some cases sync methods actually perform async operations under the hood.

However, coming from the Unix universe, it doesn't seem to hold that there's "no such thing as synchronous I/O" at least at the kernel level: I would expect .NET I/O operations to map to simple sync I/O system calls such as write and read. Also, the implementation details underlying async operations across platforms may make sync I/O more attractive at least in some cases (i.e. is async I/O implemented efficiently on Linux with mono, or with coreclr for that matter?). I admit I'm just asking questions, I don't have enough solid knowledge here.

Regardless of these questions, and I said back in #4868, there's something odd about removing such a specific sync API (SslStream.AuthenticateAsClient) for such a general "sync I/O is bad" argument, without touching any other sync API. In other words, I'd expect either all sync APIs to stay for backwards compatibility or for a more sweeping removal of sync I/O APIs.

**Comment by roji** [comment]
One more important note. You're effectively telling me to implement .NET's Open(), a sync API, by internally calling an async method and waiting on it. This is a scenario discussed (and warned against) by Stephen Toub: http://blogs.msdn.com/b/pfxteam/archive/2012/04/13/10293638.aspx

**Comment by roji** [comment]
One additional reaction to something you said before:

>> One disadvantage of async I/O is the pressure put on the thread pool...

> Pressure on a thread-pool shows poor design in the Async system. Ideally you should always try to keep the number of threads equal to the number of CPUs. Everything above that is going to add overheads of thread switching and memory consumption. This can be achieved by using TPL Async. The concurrency level can be controlled via the TaskScheduler class.

That certainly seems true in a single application that's totally within my control, but what happens in a large complex concurrent application using several 3rd-party libraries which themselves may or may not rely on the TP? It can easily become complicated to to manage this and understand the exact TP needs across everything...

At the end of the day, async APIs introduce a dependency on a global shared resource (the TP) which sync APIs don't have...

Any comments on all this?

**Comment by clrjunkie** [comment]
>That certainly seems true in a single application that's totally within my control, but what happens in a large complex concurrent application using several 3rd-party libraries which themselves may or may not rely on the TP? 

You raise a valid concern, 

Note, that in .NET the TP (e.g ThreadPool Class) actually abstracts two "Thread Pool" implementations:  A Managed Work Queue as well as  various *unrelated methods* to bind IOCP callbacks with the *Win32 Native ThreadPool*

The Socket API will invoke the callbacks using the Native ThreadPool worker thread.

If you can guarantee (which I doubt) that users of your library won't abuse your callback context with cpu bound work than an alternative approach would be to have the Socket library provide the ability to bind with a dedicated user-defined (or framework created) Native Win32 ThreadPool.

This is by the way how it's tackled in Native Win32 application development but again, the assumption here is that the usage of the dedicated ThreadPool is not abused by its caller.

I agree that having a Sync API has it's  benefits in certain situations - we don't live in a perfect world.

Furthermore, putting aside async theory, Windows scheduler is super optimized and I recall reading some time ago a benchamrk for ASP.NET that showed Sync request handling actually outperformed async/await requests handling, I didn't have a chance to investigate this therefore you should judge this yourself, try searching for it online.

**Comment by clrjunkie** [comment]
Actually, the benchmark showed sync outperforming async in a client side test which is more relevant to this issue. 

Here is the link:

http://www.ducons.com/blog/tests-and-thoughts-on-asynchronous-io-vs-multithreading

**Comment by roji** [comment]
@clrjunkie thanks for the added info and the benchmark!

@CIPop @stephentoub any comments here?

**Comment by roji** [comment]
Just wondering if there are any reactions to the above discussion...

**Comment by benaadams** [comment]
@clrjunkie that test is quite a bad implementation of both sync and async (async void with no coordination). Also 57 second to execute 1M requests is fairly slow; since the aspnet5 [Kestrel Server](https://github.com/aspnet/KestrelHttpServer) using purely async/await methods will both happily execute over 1M requests in under a second for small requests:

[![1M rps](http://aoa.blob.core.windows.net/aspnet/1million.png)](https://twitter.com/DamianEdwards/status/679743708758061056)

And crush a 10GbE network for large requests on (Azure G4 Windows VM -> Linux VM)

![](http://aoa.blob.core.windows.net/aspnet/bandwidth.png)

All using coreclr x64 and pure managed async/await and asynchronous I/O

**Comment by benaadams** [comment]
> At the end of the day, async APIs introduce a dependency on a global shared resource (the TP) which sync APIs don't have...

@roji the Task returning "async" method doesn't have to actually be async; it can just return a cached precompleted Task (like Task.CompletedTask) or [ValueTask](https://github.com/dotnet/corefx/blob/master/src/System.Threading.Tasks.Extensions/src/System/Threading/Tasks/ValueTask.cs) for returned values if it is mostly sync.

**Comment by clrjunkie** [comment]
@benaadams

>that test is quite a bad implementation of both sync and async (async void with no coordination) and they aren't comparing the same or best approaches for both.

What do you mean by “quite bad”? Have you tried to apply the best approaches to the test code to see if they make any change to the final outcome?

>Also 57 second to execute 1M requests is fairly slow; since the aspnet5 Kestrel Server using purely async/await methods will both happily execute over 1M requests in under a second for small requests:

I’m well aware of the TechEmpower Benchmark (e.g “Stress Test”) and Wrk (more than year) and suffice to say that aside from two points I’m not a big fan. I share the opinion of others that such benchmarks do not represent real world scenarios (see https://webtide.com/jetty-in-techempower-benchmarks/) Furthermore, there are many obvious points of concern for a production webserver that are not measured in this benchmark that have direct impact on performance but I currently don’t have the time to discuss them (although I think a serious discussion should take place). Saying that, there are two things I do find interesting about the stress tests 1) It tests if whether your code crashes during high load 2) it shows you what impact doing ‘new’ all over the place has and I consider the whole RPS thing as natural side effect, but looking at the bigger picture to me it’s not interesting.  Quite frankly, I’m a bit disappointed that MS decided to participate in this race as I think it’s a marketing honey pot. It’s funny that when you look into the results that start remotely to resemble a real system test with a DB, then the difference between the top runner and the "old" ASP.NET is negligible.

![tech-bench](https://cloud.githubusercontent.com/assets/15968411/12398162/803f26f0-be1a-11e5-9f7d-29e818302d25.png)



Furthermore, I think @roji main concern is about a non-pipelining request/response exchange pattern so I don’t think the wrk pipelining test result that pass 50 - 100 bytes per request back and forth is applicable here, or to the majority of uses cases for that matter.

Food for thought:

Can you *prove* that for each request generated by wrk you actually get a full valid response containing all the bits the server sent?

p.s

To avoid any doubt I don't suggest that your optimization efforts are not important. I simply don’t agree with, well with the benchmark as well as  the importance put on an RPS that changes instantly once you add a small requirement or slightly change the test environment. I simply don't think this is serious..

**Comment by benaadams** [comment]
@clrjunkie 
> What do you mean by “quite bad”?

Its very much an apples and oranges comparison. The throttling in the async was achieved by limiting the connections with ServicePoint which is not simple (see [FindConnection and SubmitRequest](http://referencesource.microsoft.com/#System/net/System/Net/ServicePoint.cs,381) when over the limits) so queuing on threadpool vs queuing on ServicePoint is what's mostly being tested here; the threadpool being far more efficient and the threadpool is what async uses for its general use.

> I’m well aware of the TechEmpower Benchmark (e.g “Stress Test”) and Wrk (more than year) and suffice to say that aside from two points I’m not a big fan.

Also my point had nothing particularly to do with the TechEmpower Benchmarks. It was that non-blocking async I/O is not slow in .NET and you aren't loosing anything in perf by comparison to the time taken by the actual I/O (similar to your raw response vs response w/ db point).

It is also hugely more scalable than (sync) blocking I/O; in fact that is one of the issue with the benchmark you linked to. It is 1M requests and the sync blocking is limited by how many requests it can process at one time as its throttled by the threadpool. One the other hand the async doesn't have that limit so it can send all the requests fairly concurrently (depending on RTT); which will probably additionally make the server unhappy; and then memory issues observed wil arise because so many responses arrive back at the same time.

> Furthermore, I think @roji main concern is about a non-pipelining request/response exchange pattern so I don’t think the wrk pipelining test result that pass 50 - 100 bytes per request back and forth is applicable here, or to the majority of uses cases for that matter.

The 12.6 GBps throughput screenshotted was from sending 300kB per request, non-pipelined, to 1000 connections which is half the page weight of for example the Angularjs or 60% of the npm homepage to pick two. So fits more with a "real world" scenario. Granted if you talk to a db that will be your bottleneck, but webserver do do a lot more than that; whether its through memory caching, serving html+img+js; streaming videos, downloading files etc

> Have you tried to apply the best approaches to the test code to see if they make any change to the final outcome?

Since I raised the criticism, I see if I can present a better approach - will get back to you.

**Comment by clrjunkie** [comment]
@benaadams

>Its very much an apples and oranges comparison. The throttling in the async was achieved by limiting the connections with ServicePoint which is not simple (see FindConnection and SubmitRequest when over the limits) so queuing on threadpool vs queuing on ServicePoint is what's mostly being tested here; the threadpool being far more efficient and the threadpool is what async uses for its general use.

Maybe.. since I didn’t write the benchmark I’m reluctant to accept various explanations. However, since the author compared two approaches using out of box framework facilities one would reasonably expect a noticeable improvement in the client side test and not “because”; thus my initial remark “putting aside async theory”. The whole point of the link was to bring attention to the fact that an async api by itself says nothing about performance if the underlying implementation doesn’t deliver upon it.

>Also my point had nothing particularly to do with the TechEmpower Benchmarks.
 
Ye’h but it kinda slipped in…  The 1M rps you mentioned was linked to Damian post and as far as I understand he uses or is inspired by the TechEmpower scenarios. Know any different for this particular test? https://github.com/aspnet/benchmarks 

>It was that non-blocking async I/O is not slow in .NET and you aren't loosing anything in perf by comparison to the time taken by the actual I/O (similar to your raw response vs response w/ db point).

I didn’t say nor did I suggest that you are losing anything in perf by comparison to the time taken by the actual I/O. The players here are: .NET Async support, which depends on compiler generated code, threadpool implementation, o/s signaling implementation Vs. Sync support, which includes use of the Thread library and the O/S scheduler – a facility that is heavily optimized for fast context switching between many threads.

> It is also hugely more scalable than (sync) blocking I/O; in fact that is one of the issue with the benchmark you linked to. It is 1M requests and the sync blocking is limited by how many requests it can process at one time as its throttled by the threadpool. One the other hand the async doesn't have that limit so it can send all the requests fairly concurrently (depending on RTT); which will probably additionally make the server unhappy; and then memory issues observed wil arise because so many responses arrive back at the same time.

See first response above.

> The 12.6 GBps throughput screenshotted was from sending 300kB per request, non-pipelined, to 1000 connections which is half the page weight of for example the Angularjs or 60% of the npm homepage to pick two. So fits more with a "real world" scenario. 
Nothing is “real world” with a burst test on pre setup connections, unless you consider DDOS the common scenario.

The 12.6 GBps figure doesn’t say much as it aggregates protocol level headers etc.. All I can conclude from this figure by doing a ridiculously naïve calculation is that the server processed 44040/rps of such requests which is 44/rps per connection (1KB = 1024B). What does this have to do with a Sync? Did you see if a server backed by a 1000 thread threadpool shows substantially lower earth shattering results?

Loading Angularjs front page generates 53 distinct requests on EDGE empty cache, even it reuses Keep-Alive connection I don’t see what’s the takeaway or how you correlate it to your test. 

>Granted if you talk to a db that will be your bottleneck, but webserver generally does a lot more than that; whether its through memory caching, serving html+img+js; streaming videos, downloading files etc

But it’s generally on the path… unless you’re after “geocities” style websites (gosh I miss those :)
Caching is externalized in any serious high traffic site. I would probably stick with IIS (or NGINX) for kernel level caching and file transfer without passing through user-mode or more probably a CDN that employs dedicated solutions for streaming videos / heavy file downloads.

Anyway I think this whole performance discussion is completely deviating from the main concern of this issue which is single threadpool pressure.

**Comment by benaadams** [comment]
> Did you see if a server backed by a 1000 thread threadpool shows substantially lower earth shattering results?

Can't currently manually increase the threadpool in coreclr https://github.com/dotnet/coreclr/issues/1606 & https://github.com/dotnet/cli/issues/889 and if you try 1000 connections the server out of the box it crawls. Obviously there aren't 1000 thread pool threads to start with; would have to jam the server for 16 min to create them.

**Comment by clrjunkie** [comment]
The thread pool would obviously lazy create them, *when needed*. The point was to guarantee a theoretical upper bound for processing.  thread != connection, final outcome matters.

**Comment by benaadams** [comment]
Yeah, and it dies; because the I/O is inherently async and the sync-blocked threadpool threads use up all the threadpool threads in a blocked waiting state so the async I/O has nothing to run on to notify them to unblock.

**Comment by benaadams** [comment]
The full framework uses two threadpools to work round the issue

**Comment by clrjunkie** [comment]
@benaadams

>Yeah, and it dies; because the I/O is inherently async and the sync-blocked threadpool threads use up all the threadpool threads in a blocked waiting state so the async I/O has nothing to run on to notify them to unblock.

Implementation dependent. Look at other FW samples.

**Comment by clrjunkie** [comment]
@benaadams

>The full framework uses two threadpools to work round the issue

Read my post addressing this.. above, above, above...

**Comment by sqmgh** [comment]
>However, since the author compared two approaches using out of box framework facilities one would reasonably expect a noticeable improvement in the client side test and not “because”; thus my initial remark “putting aside async theory”.

Did we read the same article?

Depending on where in the article you are, the sync method doesn't actually use an out of the box method. They async method might not either; It's hard to tell because the initial implementation apparently never returned any requests. There is text talking about changing the implementation in an attempt try and get more performance, and some alluding to using less 'out-of the box' methods to do so. However, it's never noted what exactly was changed, or what implementation was actually used in further tests. The only thing that can be said for sure is that the async code that is provided is very much not best practice. Last I saw` async void` will actually cause Kestrel to crash.

Yes, the initial example code does use an 'out of the box' threading method. However, the author says that the initial example does such little work on each request that it hardly uses any threads.
Later in the article when the author modified the benchmark to do _actual_ work, he had to write his own 'custom implemented parallel execution engine'. Which he eventually limited to 500 threads for 5000 total requests. 500 threads that weren't enough to keep up with the async implementation, and causes him to conclude that async gives better performance when doing actual work. If you go by the server side section, much much better.

Providing this benchmark as supporting evidence of something, and then immediately saying the techempower benchmarks are meaningless because they aren't "real world" is extremely strange to me.
The initial benchmark is basically the wrk end of the techempower benchmark. Since the response that it's receiving is only "Hello world!" it's technically doing _even less_ than the plaintext techempower benchmark is.
So with that logic, and ignoring the potentially dubious implementation, I think the conclusion would be that sync might be better is unrealistic microbenchmarks. However, since that's unrealistic and web servers tend actually do work like talking to databases, then everything should be async anyways.

**Comment by clrjunkie** [comment]
clrjunkie wrote:

>Furthermore, putting aside async theory, Windows scheduler is super optimized and I recall reading some time ago a benchamrk for ASP.NET that showed Sync request handling actually outperformed async/await requests handling, I didn't have a chance to investigate this therefore you should judge this yourself, try searching for it online.

---

>Actually, the benchmark showed sync outperforming async in a client side test which is more relevant to this issue.

Here is the link:

http://www.ducons.com/blog/tests-and-thoughts-on-asynchronous-io-vs-multithreading

---

>Maybe.. since I didn’t write the benchmark I’m reluctant to accept various explanations. However, since the author compared two approaches using out of box framework facilities one would reasonably expect a noticeable improvement in the client side test and not “because”; thus my initial remark “putting aside async theory”. The whole point of the link was to bring attention to the fact that an async api by itself says nothing about performance if the underlying implementation doesn’t deliver upon it.

**Comment by roji** [comment]
Am back from a long project, sorry for not having participated more in the conversation above.

First of all, it's somewhat disappointing to not hear anything from anyone at Microsoft on this (e.g. @CIPop).

Beyond the question of "pure async vs. sync performance" discussed above, there are other factors as I tried to say. Thread pool pressure is one of them - I may want the added predictability/determinism that sync calls provide in that they don't add a dependency on the thread pool, which may be or may not be a problem in a given application. It's a choice that should be left to the programmer rather than saying "sync I/O is always bad".

But more importantly, one of the key points of CoreCLR is portability. Saying "sync I/O is always bad" means you're sure that across Linux, MacOS etc. the kernel's sync system calls offer no advantage over whatever async mechanism the runtime does - that seems false, and should at least be shown.

**Comment by CIPop** [comment]
[Network Programming for Microsoft Windows 2nd Ed.](http://www.amazon.com/Network-Programming-Microsoft-Windows-Edition/dp/0735615799), Table 6-3 has a comparison between all I/O models available in Windows at that time. Their conclusion was that IOCP was the best I/O model at that time while blocking calls was performing the worst.

I don't know of any architecture (hardware or software) that implements synchronous I/O when the speeds between the processing unit and the transceiver are orders of magnitude apart. If that would be true, the processing unit's computing power is wasted by either spin-waiting or powering off while waiting for the transceiver to complete the transfer.

For our discussion, the CPU is orders of magnitude faster than the network card and other processes exist in the system. There is no way to handle sync NIC/memory transfers unless the CPU waits. Transfers are managed by DMA which signals the CPU via interrupts whenever the transfer has been completed. The apparent blocking syscalls for *NIX are handled by suspending execution threads (processes) until the I/O is completed. Other platforms such as NodeJS use systems such like event loops/message pumps to dispatch completions on a single thread (this is similar in architecture with the WSAAsyncSelect).

In Windows, a thread comes with high overheads for the CPU and memory resources. The performance guideline is to have the of threads equal to the number of CPUs and no threads in waiting/sleep states. 
In other words: apps should always use threads (CPUs) for compute-bound operations and perform I/O operations asynchronously.

**Comment by roji** [comment]
@CIPop thanks for the reply.

I understand (and am aware of) what you're saying. Obviously sync I/O doesn't imply having the CPU spin-wait or power off - the thread/process is simply put into a non-runnable state until the operation completes. So the question here is not how I/O is actually handled at the lowest levels, but rather how the interaction between the application and the kernel takes place with regards to I/O.

In that context, it may make sense in certain scenarios to have independence from the thread-pool which sync I/O calls provide - you know which thread will execute "the callback", you're 100% sure it will be there waiting when the operation completes. You don't depend on all kinds of external conditions like what other libraries are operating inside the application, putting pressure on the thread pool and making things less predictable for you.

In addition, you need to take into account non-Windows platforms. AFAIK thread overheads in Linux aren't as high as in Windows, and async I/O (both underlying kernel mechanisms and implementation in .NET runtimes) isn't necessarily as widely-used or efficient as Windows. The "everything async" vs. "let's leave room for sync" debate has taken place before, and despite some current trends (and the hype around NodeJS) in some scenarios sync I/O may make sense.

Don't get me wrong, I agree that in the vast majority of cases I/O should be async. I'm just arguing that you should be leaving a choice to programmers rather than completely remove the possibility of sync I/O.

**Comment by roji** [comment]
@CIpop and others, I'll just give a quick concrete example. I'm the maintainer of Npgsql, the PostgreSQL driver for .NET. We've frequently received complaints that under load, and frequently at program startup, DNS queries performed by Npgsql time out. Now, the current stable Npgsql does DNS resolution with `Dns.{Begin/End}GetHostAddresses`. What seems to be happening, is that when the program starts the thread pool has relatively few threads. If a large amount of requests come in and a lot of async operations are in progress, the thread pool starts allocating new threads - but this process takes time; and during these time it would seem that the operation times out.

I'm mentioning this specific scenario because in CoreCLR, the sync API `Dns.GetHostAddresses` has been dropped, much like `SslStream.AuthenticateAsClient`, because "sync API is always bad".

**Comment by benaadams** [comment]
@roji I think that's a bug, see https://github.com/dotnet/corefx/issues/8371

**Comment by roji** [comment]
@benaadams see my comment https://github.com/dotnet/corefx/issues/8371#issuecomment-217705749: the risk of threadpool starvation seems like a general problem with async, not a specific problem with how DNS resolution works - that's the point I'm trying to make here (unless I'm missing something).

**Comment by GSPP** [comment]
The starvation issue should always be worse with sync IO running on pool threads as it is with async completion callbacks posted to the TP.

It is true that "pathologically" bad cases can happen due to scheduling etc. but the way to avoid *all* of those systematically is to not overload the TP.

In a pure async IO workload the only way the pool can be overloaded is when the CPU becomes saturated. That case is to be avoided as well, and it's just as troublesome with many threads performing sync IO.

> What seems to be happening, is that when the program starts the thread pool has relatively few threads.

[Yes, this is a significant problem with the .NET thread pool.](https://github.com/dotnet/coreclr/issues/1754) I really wish this was addressed.

In the Dns case that you mention the root cause is the bad async implementation, not async IO (there is no async IO ATM).

> sync API Dns.GetHostAddresses has been dropped, much like SslStream.AuthenticateAsClient, because "sync API is always bad".

I did not know that CoreClr has a strategy of dropping synchronous IO methods?! That would be a really bad policy. Async IO has a significant productivity cost and it is a bad default choice.

Async IO should be used when a tangible benefit can be obtained by doing that. Such benefits are saving threads and GUI programming simplification. If enough threads are available saving threads has zero customer value and negative value for productivity. A common misconception is about performance, so note that DNS queries do not become one bit faster by using async IO.

**Comment by roji** [comment]
@GSPP we more or less agree...

About dropping synchronous IO methods, see https://github.com/dotnet/corefx/pull/4868 where I complain about the removal of `SslStream.AuthenticateAsClient()`. But closer to our subject, netstandard1_3 drops the sync `Dns.GetHostAddresses()` - you have to use `Dns.GetHostAddressesAsync()` (which ironically isn't async as in #8371!!). So there's seems to be a "let's kill all sync I/O" argument in the design of CoreCLR...

However I don't agree 100% with the starvation issue always being worse with sync I/O. If you know your precise application needs and have no serious scability goals (as you say) a sync I/O program can be more predictable and less sensitive to issues like #1754. You own the threads, you have no dependency on the TP whatsoever, everything is simple.

But that's not very important - my only goal here is to make a case for leaving sync I/O methods in .NET...

**Comment by GSPP** [comment]
Maybe you should open a ticket dedicated to that request. This one is a discussion about many things. Make a formal request to continue to provide synchronous methods except in special cases.

Fyi, starting with .NET 4.6 (I think) the ADO.NET query methods are internally async and the sync methods are sync-over-async. Not sure if that was a good choice. Might be an OK engineering trade-off. On the other hand database calls are a classic case where async is *almost always* not helping throughput.

**Comment by roji** [comment]
@GSPP this is exactly what this issue is for (#5044)... But nobody from Microsoft seems to be taking it seriously.

Thanks for mentioning about .NET 4.6, I didn't know that! Can you point to any issue/PR where this is discussed? Any idea what's the status with CoreCLR?

**Comment by roji** [comment]
Also, why do you say async isn't useful for database calls? It's I/O like any other, why not make your webapp be more scalable by not blocking a thread while waiting for a database response?

**Comment by GSPP** [comment]
If threads are not scarce the scalability gains are zero.

DB calls are quick, almost by construction. If they are slow then the database is overloaded. The DB becomes overloaded far quicker than the web app loses its ability to provision threads.

In reporting workloads queries take a long time but there are few of them at the same time.

The number of threads consumed is equal to (latency * frequency). So 10ms query latency and 1000 queries per second (a lot!) cost you 10 threads on average which is nothing to speak of.

100s of threads under ASP.NET are not an issue, but that would already be an extreme case.

**Comment by roji** [comment]
@GSPP interesting argument. Although nowadays databases are sometimes far away geographically (e.g.another datacenter) so a lot of the latency could be network rather than database load. But I understand the general argument.

**Comment by benaadams** [comment]
The async "cost" is between 120ns and 10ns https://github.com/ljw1004/roslyn/blob/features/async-return/docs/specs/feature%20-%20arbitrary%20async%20returns.md

A blocking sync call is a syscall to thread de-schedule and then a reschedule when complete, not sure of latencies

**Comment by roji** [comment]
@benaadams in most cases there's also a memory allocation cost for the Task being returned (in some cases you can return cached instances but not always).

**Comment by GSPP** [comment]
@benaadams the descheduling happens with async as well iff there is not another TP item to work on. Unless the CPU is *very* loaded (like 90%+) the chance is very high that there is none.

Async improves throughput at very high CPU loads indeed. At lower loads it increases CPU consumption. Don't generally think of async as a way to save CPU.

Also it *enables* high levels of IO parallelism that otherwise would be thread constrained.

**Comment by roji** [comment]
Oh interesting, I didn't know about ValueTask!

**Comment by JeffCyr** [comment]
Here's my thoughts on the matter.

There is a valid point brought by @roji, in large and complex applications, the .Net ThreadPool is often misused by doing synchronous IO in it. This makes relying on the ThreadPool risky.

However, the root cause of the ThreadPool being misused is because the synchronous IO methods exists in the first place. If all IO methods were asynchronous, then it would be explicit that a method may be doing IO. Then the caller would be more inclined to await the returned task instead of blocking on it.

So I think that .Net Core is taking the right stance by dropping all synchronous IO methods.

**Comment by GSPP** [comment]
@JeffCyr your argumentation should discuss the downsides of async as well. If there were none everyone would agree.

**Comment by roji** [comment]
@JeffCyr I/O isn't the only reason for a method to block. In another words, even if you get rid of all the sync I/O methods, users can still start some intense computational task and block a thread.

And like @JeffCyr says, there are downsides to async, e.g. async methods usually imply increased heap allocations (because of Task instances). All I'm saying is leave the choice up to the programmer.

**Comment by clrjunkie** [comment]
@roji 
Let me give you a friendly advice: 

Drop it until you can show a compelling business scenario for including I/O API’s that are not *Task based*.

Async API’s have one and only one technical goal: reduce the amount of context switches that would otherwise occur due to large amount of running threads servicing submission and completion of I/O request.

What’s “large” is highly subjective and scenario driven and it’s a waste of time to reason about the implications using hypothetical examples or micro optimizations. In other words, for every reasonable argument I can give a counter reasonable argument.

Async is not about fairness, and it doesn’t come without its cost, but it turns out there are  common scenarios where the benefit exceeds the cost – but not always. The only sensible question to ask is whether Async API’s benefits YOU.

There are plenty of valid reason I can think of why Microsoft has chosen to go solely with Task-based I/O API’s that have little to do with high performance, for example:

1. There are no noticeable differences between the two for the majority of customer use cases so it’s cheaper to implement, document, and support one version.

2. Less arbitrary threads means you can squeeze more VM’s on average to a single physical host and by that reduce cost. (for both MS and customers)

3. Web Server are less susceptible to malicious slow network client attacks.

4. It’s easier to write await than deal with APM so the chance that an app/Winform developer will block the UI thread is reduced by virtue of not having the possibility to call sync API’s to begin with leading to better quality.

5. Some would appreciate the lack of choice.

6.Async sounds cool.

So again, putting it simply instead of getting frustrated ask yourself: would sync API’s really provide substantial benefit to the majority of my library users?  (Is it worth the cost for you and Microsoft?)

Don’t get me wrong with this “biz-talk” I have total sympathy to your concerns, Heck I’m now investing in a completely different framework/language because it’s important to me to use HTTP components that are purely written down to the socket level in the language of my choice, but that’s *my personal requirement* which may very well not be applicable to others and I have come to terms with that.

**Comment by clrjunkie** [comment]
and stack size (implied)

**Comment by roji** [comment]
@clrjunkie,

> Drop it until you can show a compelling business scenario for including I/O API’s that are not Task based.

This is a discussion about programming technicalities, I'm not sure what "a compelling business scenario" would consist of here, but I'll sum up again some disadvantages of async:

* Considerably more complicated programming model because of the thread switching - programmers have to be careful with the synchronization context, possibilities of deadlock. The "business" scenario of sync I/O is simplified programming
* Increased heap allocations (Task)
* Less predictable/determinstic in large applications because of the dependency on the TP

> What’s “large” is highly subjective and scenario driven and it’s a waste of time to reason about the implications using hypothetical examples or micro optimizations. In other words, for every reasonable argument I can give a counter reasonable argument.

I'm not sure what hypothetical examples or even less, micro optimizations I've referred to...

> Async is not about fairness, and it doesn’t come without its cost, but it turns out there are common scenarios where the benefit exceeds the cost – but not always. The only sensible question to ask is whether Async API’s benefits YOU.

I'm confused now... I'm not saying async is bad - just that in some scenarios it's not the right choice. When you're saying that "there are some common scenarios where the benefit exceeds the cost - *but not always*", isn't that an argument for leaving the sync I/O option?

> There are no noticeable differences between the two for the majority of customer use cases so it’s cheaper to implement, document, and support one version.

.NET is a framework, it's support to support the needs of a wide base of users. There are a lot of edge cases covered by .NET that don't meet "the majority of customer use cases" and that's a good thing. I sense something else happening here - there's a refusal to admit that async does have its issues, and that there *are* cases where sync I/O makes more sense.

> Less arbitrary threads means you can squeeze more VM’s on average to a single physical host and by that reduce cost. (for both MS and customers)

I have no idea what that means, are TP threads somehow less arbitrary in some way?

> Some would appreciate the lack of choice.

Really... maybe they would, I definitely won't. But I guess it's a question of ideology.

> So again, putting it simply instead of getting frustrated ask yourself: would sync API’s really provide substantial benefit to the majority of my library users? (Is it worth the cost for you and Microsoft?)

Let me simplify things here. There's only one relevant question here: are there scenarios where sync I/O makes more sense than async I/O? If you answer "no" that's one thing - I think you're wrong but we can discuss. If you answer "yes but the majority don't care" then we should start taking a look at all the entire .NET API surface and removing rarely used APIs.

**Comment by benaadams** [comment]
> > Less arbitrary threads means you can squeeze more VM’s on average to a single physical host and by that reduce cost. (for both MS and customers)

> I have no idea what that means, are TP threads somehow less arbitrary in some way?

I assuming is referring to a Thread min size 1MB and Task min size 80 Bytes; also N tasks to the O/S can be 1 thread, but N threads is N threads so less work context switching and management for the O/S?

**Comment by clrjunkie** [comment]
@roji 
>Considerably more complicated programming model because of the thread switching - programmers have to be careful with the synchronization context

Prove it. Do you see the Internet/forums flooded with questions about this.

>Increased heap allocations (Task)

So what?

>Less predictable/determinstic in large applications because of the dependency on the TP

For who?

>When you're saying that "there are some common scenarios where the benefit exceeds the cost - but not always", isn't that an argument for leaving the sync I/O option?

No, because I didn't include a $ nor an enterprise customer/scenario.

>there's a refusal to admit that async does have its issues, and that there are cases where sync I/O
 makes more sense.

Don't think so.

>I have no idea what that means, are TP threads somehow less arbitrary in some way?

See @benaadams  response

**Comment by roji** [comment]
Uhh, I'm a bit lost now but I'll try to answer these in good faith.

>> Considerably more complicated programming model because of the thread switching - programmers have to be careful with the synchronization context

> Prove it. Do you see the Internet/forums flooded with questions about this.

One issue that seems to bite people all the time is [blocking on async](http://blog.stephencleary.com/2012/07/dont-block-on-async-code.html). I'm not going to start finding popularity metrics to prove to you that this is a serious issue, but are you seriously claiming there's no added complexity because of SynchronizationContext, ConfigureAwait, the thread locality and all the rest of it? Do you seriously consider async to be just a simple, entry-level language feature that doesn't require anything more from the programmer than understanding sync I/O?

>> Increased heap allocations (Task)

> So what?

I don't know how to answer this one. C# is a language which includes unsafe, allowing you to do pointer operations for performance. We're not in Java-land where we just throw performance out the window. If you don't understand the implications of heap allocations and GC pressure go read up on it.

>> When you're saying that "there are some common scenarios where the benefit exceeds the cost - but not always", isn't that an argument for leaving the sync I/O option?

>No, because I didn't include a $ nor an enterprise customer/scenario.

That's simply dishonest. Either there are scenarios where "the benefit exceeds the cost", or there aren't. If they are, that cost translates into $ or "$ and enterprise customer/scenario".  We're programmers, we're debating a technical issue., Programmer productivity (and even heap allocations) are important.

>> there's a refusal to admit that async does have its issues, and that there are cases where sync I/O makes more sense.

> Don't think so.

Uh, ok. You know you can't just decide that the entire world shouldn't care about increased heap allocations, right? There are people with different needs out there. You seem to be closing your eyes to any sort of objection to using async?

>> I have no idea what that means, are TP threads somehow less arbitrary in some way?

> See @benaadams response

Frankly, I'm going to let this one go because the other arguments are much more important (I think there's just some confusion here).

**Comment by JeffCyr** [comment]
I get an impression that this discussion won`t come to an agreement, but here it is for the sake of the argument:

>@JeffCyr I/O isn't the only reason for a method to block. In another words, even if you get rid of all the sync I/O methods, users can still start some intense computational task and block a thread.

This can be mitigated by spawning the long computational task with LongRunning option. But to be fair, from experience almost every ThreadPool starvation I debugged were caused by blocking IO and not cpu bound tasks.

>And like @JeffCyr says, there are downsides to async, e.g. async methods usually imply increased heap allocations (because of Task instances). All I'm saying is leave the choice up to the programmer.

Have you measured the impact of this "increased" heap allocations? The async methods return cached completed tasks for the most common result value when completed synchronously, and even when a Task is created, it won't live past the gen0 collection which is very cheap to collect. When the operation is truly async, yes a task will be created that might make its way to gen2, but it will still be cheaper than creating/destroying a thread and keeping the stack size in memory while waiting for the blocking operation.

As for blocking on async code, while the rule of thumbs is to call it async, there is nothing wrong in blocking on async code if all await instruction in it use ConfigureAwait(false) which the .Net framework should always do. The performance impact is also negligible, many blocking IO operation are implemented in the OS by blocking the asynchronous version anyway.

**Comment by roji** [comment]
@JeffCyr I agree we're probably not going to agree. For what it's worth some comments on what you say specifically.

>>@JeffCyr I/O isn't the only reason for a method to block. In another words, even if you get rid of all the sync I/O methods, users can still start some intense computational task and block a thread.
>This can be mitigated by spawning the long computational task with LongRunning option. But to be fair, from experience almost every ThreadPool starvation I debugged were caused by blocking IO and not cpu bound tasks.

Sure this kind of problem can be mitigated - whether it's computational or sync I/O. The point is that the problem is there, programmers have to deal with them, it's a complication in the programming model which you don't have in sync. Whether it's a huge problem or not is another issue.

>> And like @JeffCyr says, there are downsides to async, e.g. async methods usually imply increased heap allocations (because of Task instances). All I'm saying is leave the choice up to the programmer.
> Have you measured the impact of this "increased" heap allocations? The async methods return cached completed tasks for the most common result value when completed synchronously, and even when a Task is created, it won't live past the gen0 collection which is very cheap to collect. When the operation is truly async, yes a task will be created that might make its way to gen2, but it will still be cheaper than creating/destroying a thread and keeping the stack size in memory while waiting for the blocking operation.

I can't measure the impact of increased heap allocation because, like everything else, it's going to depend on your application. GC pressure caused by spurious allocations are a major source of performance issues in general, enough that [System.Xml itself internally uses a non-heap Task alternative](https://github.com/ljw1004/roslyn/blob/features/async-return/docs/specs/feature%20-%20arbitrary%20async%20returns.md) (thanks @benaadams). Whether a Task instance will be cached (because it contains a common result) or whether it will make it out of gen0 is again, going to depend a lot on your application - I generally try to avoid statements like "it won't live past the gen0 collection".

Nobody has been talking about creating or destroying threads here - just one, simple thread blocking on I/O.

> As for blocking on async code, while the rule of thumbs is to call it async, there is nothing wrong in blocking on async code if all await instruction in it use ConfigureAwait(false) which the .Net framework should always do.

That's true, but it's another complication which the programmer needs to understand with async. You guys are missing the point; I know how to avoid these deadlocks and performance problems, the solutions are there. It's just that programmers need to deal with these issues in async, and not in sync. That is a cost of the async model.

> many blocking IO operation are implemented in the OS by blocking the asynchronous version anyway.

I'm not an expert on Windows kernel APIs, but I do know that at least in Linux sync I/O APIs are cleanly mapped to sync system calls which have been there for decades.

At the end of the day you're advocating removing a standard API that exists pretty much in every language that every existed. I'm just saying the choice should be left to the user, depending on their app.

**Comment by JeffCyr** [comment]
>Nobody has been talking about creating or destroying threads here - just one, simple thread blocking on I/O.
I was referring to the one thread per request model, in this case yes the overhead of creating/destroying thread and keeping the stack size in memory will be greater than the "heap pressure" caused by async.

If you are talking about a single threaded application doing blocking IO, then I doubt it's an application with high CPU performance requirement, it's more about an API experience.

>At the end of the day you're advocating removing a standard API that exists pretty much in every language that every existed. I'm just saying the choice should be left to the user, depending on their app.

I don't see any golang or nodejs programmers complaining about the missing blocking IO calls. .Net Core is a new platform and it chose to be an async IO platform. If you truly want blocking IO calls, then .Net Core may not be the platform you are looking for.

By the way, while this discussion is interesting, the end result is quite futile. The Async IO choice in .Net Core have been made since its inception, Microsoft/.Net Foundation won't change their mind about this decision now.

**Comment by roji** [comment]
>> Nobody has been talking about creating or destroying threads here - just one, simple thread blocking on I/O.
> I was referring to the one thread per request model, in this case yes the overhead of creating/destroying thread and keeping the stack size in memory will be greater than the "heap pressure" caused by async.

If you are talking about a single threaded application doing blocking IO, then I doubt it's an application with high CPU performance requirement, it's more about an API experience.

Why on earth can't a single-threaded application with block I/O have high CPU performance requirements? And my application can have multiple threads which I manage manually as I see fit...

>> At the end of the day you're advocating removing a standard API that exists pretty much in every language that every existed. I'm just saying the choice should be left to the user, depending on their app.

> I don't see any golang or nodejs programmers complaining about the missing blocking IO calls. 

So what? nodejs is primarily a web development language whereas C#/.NET has a much wider scope. I don't think you'll find nodjs programmers complaining about the lack of unsafe pointer support, and yet C# includes them - do you think that should be removed as well because nodejs people aren't complaining?

> .Net Core is a new platform and it chose to be an async IO platform. If you truly want blocking IO calls, then .Net Core may not be the platform you are looking for.
> By the way, while this discussion is interesting, the end result is quite futile. The Async IO choice in .Net Core have been made since its inception, Microsoft/.Net Foundation won't change their mind about this decision now.

Oh really? Can you point me to a place where .NET Core declares this in any official way? Even more important, can you explain *why most sync I/O operations have been retained* (e.g. `Stream.Write()`)?

What seems to be happening here, is that no official decision has actually taken place, but in some specific areas of the API (e.g. Dns, SSLStream) some programmers inside Microsoft have decided to drop sync. This isn't part of a general thing in CoreCLR.

**Comment by roji** [comment]
@JeffCyr, are you an employee of Microsoft in some way, stating that "Microsoft/.Net Foundation won't change their mind about this decision now"?

**Comment by JeffCyr** [comment]
>@JeffCyr, are you an employee of Microsoft in some way, stating that "Microsoft/.Net Foundation won't change their mind about this decision now"?

I am not, I might have overstated this last affirmation, this is speculation on my part.

**Comment by aL3891** [comment]
Is the question here really about sync vs async though? isn't it more about the weight of the Tpl/threadpool model? I do think Gc pressure and thread/thread pool starvation are very valid concerns, but I would prefer a more flexible solution that would allow me to write a single thread, synchronous version (using manualresetevent or whatever) if I wanted too, but also to do something else like my own thread pool with dedicated threads. 

Then again if you really _really_ want to maximize performance, maybe dropping to the OS specific apis are the way to go anyway, but it would be really neat to have a lower overhead option for async calls in general  as well. Maybe just being able to pass in your own task scheduler to all *Async() calls would be a nice middle ground (I know you can do this in a lot of places already)

The main reason as I see it for the synchronous versions is convenience, but personally I find the `.Result` to be simple enough :) I do think you're right in voicing your desire for the other sync methods to be added though, the corefx does listen and if enough people want it, i'm sure they'll add it.

**Comment by damageboy** [comment]
Hi All, I just found this thread and my blood pressure is already starting to rise

I see that in #4868, @roji was told by @CIPop that:
> Blocking threads on I/O is always a bad programming pattern. To promote good patterns we are trying to remove all sync I/O public APIs.

Later in that thread he was asked to open a separate issue, which he did here, where from what I can tell, I can mostly see different opinions about .NET TP/GC/async/sync being exchanged.

As far as I'm concerned, if @CIPop is not messing with us, he's basically saying what amounts to: _"Oh, by the way, we, here at MS and the .NET Foundation, took some hallucinatory drugs and decided that the hell with 40+ years of synchronous I/O APIs, we're dropping it_" **<mic-drop>**

Is this real?
Can I see a real discussion with the community, initiated by MS/someone in the CLR/CoreFX teams that actually says: "Look, we're _thinking_ about dropping sync I/O APIs, what are your thoughts?"
Or is this again the resurgence of the "We know better than all of you approach" that we are supposed to be supposedly over with?!?

Can someone from MS REALLY get on this and state you intentions CLEARLY about sync I/O.
Not as a half-assed comment on a missing method in `SslSteam`, but seriously approach to community and **discuss** (which I've seen very little of) with IT the pro and cons of what you are thinking of doing, AFTER you've actually declared the scope of the intended change...

In case it wasn't clear that I would like some "official" response: What are you planning to do? Is this a half-assed attempt to remove sync I/O APIs in corefx parts where the sun doesn't shine like `SslStream`, or have you completely lost it and are actually intending to drop `Stream.Write()/Read()` because you drank some of nodejs's kool-aid?

What I see here is an active .NET developer which was told to open a discussion about a global, earth shattering change in .NET to remove sync APIs and is getting zero attention as to the claims/questions he is raising from the same people who asked him to open said discussion and others in Microsoft

**Comment by JeffCyr** [comment]
@damageboy The Stream abstraction is not IO bound, e.g. MemoryStream, so I don't think Stream.Read/Write will go away.

**Comment by damageboy** [comment]
`<cynical>`
@JeffCyr ah, wonderful, so by induction I can do SslStream over MemoryStream, so SslStream is not I/O bound as well, thanks for clearing that up, so I assume that MS is already in the process of adding SslStream sync support, thanks @CIPop !,
Case closed?
`</cynical>`

No offense @JeffCyr, but I would like to hear some response from someone in MS that can actually speak about their entire direction with the duality of sync/async

There's only so much that we, the community, can discuss amongst ourselves with no back-and-forth with MS while some MS devs make wild claims about dropping sync-I/O because "it's bad, don't you know? Someone wrote a blog post about it".

**Comment by damageboy** [comment]
@JeffCyr btw, you did, in passing, make me think of a coming trend in I/O HW systems, where I/O is actually becoming so fast, that the O/S is increasingly becoming a bottleneck when it comes to I/O performance.

More and more, we are seeing new low-latency I/O components, be it network cards that support user-space (e.g. non-blocking) networking, memory-mapped non-volatile storage (here's a recent BUILD talk about RAM-like storage systems: https://channel9.msdn.com/Events/Build/2016/P466)

It is entirely possible that as the H/W landscape continues to rapidly change, for certain types of I/O async might become too much of an overhead when compared to the fast-path sync option of future H/W...

That is why any such claim as "async is better, no matter what the underlying I/O is" is laughable at best.

I would like to have a serious discussion between whomever is capable of representing the current thought-process of MS is and us, the community...

**Comment by GSPP** [comment]
> Blocking threads on I/O is always a bad programming pattern. To promote good patterns we are trying to remove all sync I/O public APIs.

Yes, this would be a bad policy. It is important to understand how much trouble async APIs are causing. Stack Overflow is full of C# questions where someone had an async problem. The Roslyn commit logs regularly have bugs caused by async programming patterns.

Async is very costly in terms of developer time! It is a complicated trade-off and across the board rules such as "always use X" are not possible.

The .NET ecosystem is still worshiping the async fad. It was much worse 2 years ago. Finally, opinions are getting more nuanced and I hope that the sync/async trade off will be understood more widely in the future.

**Comment by roji** [comment]
@damageboy and @GSPP thank you for your voices in this, I agree 100%.

**Comment by aL3891** [comment]
I understand that you're frustrated, but you should know that in my experience the teams generally do not engage in discussions where people call them names (I lurk quite a bit on the dnf repos). 

If you want them to engage in a discussion you have to give them more than _you've just got nodejs envy and we've done it like this for 40 years_. Remember that the whole idea with .net core is to trim the fat and be modular, they _want_ to be restrictive with what they add because once its in there, you cant take it out. You have to provide really compelling arguments to why this should be _in the framework_ and not in say, an external nuget package as extension methods. _What specifically makes async (with `.Result`) so much more difficult to use compared to a sync call?_ _What other drawbacks does the Tpl async model have?_ Allocations and starvation are great examples that's been given already. 

@CIPop is a microsoftie so while it might be short of a completely official word from Microsoft I think his views here pretty much reflect the others in corefx teams (correct me if I'm wrong though)

I'm not trying to be a jerk or shut you up, I just think the way you're approaching the issue is not the most productive way to get your way. I do agree that the response from the team has been limited, but lets also remember where they are/where in the release cycle when this issue was raised. I can absolutely see something like this begin added post rtm if enough compelling reasons are given and enough people want it

**Comment by damageboy** [comment]
@aL3891 I do understand your point, but the parent issue / discussion was "opened" / reported back in December 16th, 2015, since then it was addressed ONCE, by a microsoftie, without really stating the global plans for async/sync I/O APIs after clearly stating, "oh yeah, we're basically dropping sync", here it is again, read it, and imagine you're me, reading this out of the blue, 3 hours ago:
> Blocking threads on I/O is always a bad programming pattern. To promote good patterns we are trying to remove all sync I/O public APIs.

The only thing said in that response (19 days ago) didn't address the previous comment, quoted a table from a windows-only networking book from 2002, concluded that I/O will always be slower by orders of magnitude when compared to CPU and left it at that.

I have a long list of objections to many of the things stated there, but before I list them, I would still like to get some response from MS/corefx people as to what their plans are, as one of their team members laid out a claim that they are "removing public sync I/O APIs" without providing any firm details.

I DO understand the copious amounts of pressure the team(s) must be under to deliver a working first release, but it's been almost 6 months with one response, from one corefx team member, that HE asked to OP to open in the first place, to a 60 comment issue/thread filled with non-MS people hyperbole-ing about what the holy spirit of MS means by "removing sync I/O APIs" and why...

Is this (me) being a jerk given the lack of interaction / clarity over 6 months?
Waiting patiently sure didn't help, what/how would you suggest I do, as the original comment obviously pisses me off?

**Comment by aL3891** [comment]
Hey, as you say i'm just an outsider, as we all are, and i'm just guessing. As I said i think you and everyone else should speak their mind, imo a frank discussion is always better than no discussion, but that's me. I've had the most luck pinging various people on the team or raising new, more specific issues. Right now it covers at least three things:
 
* *Sync apis vs async when it comes to performance* starvation/heap allocation, predictability of the thread pool vs os thread scheduling 
* *Ease of use of async methods* compared to sync methods, is the additional syntax/considerations worth it? Should corefx be opinionated and push users towards what they consider correct?
* *The consistency of sync apis* Some classes have them, others don't, is this intentional?

Each of these can probably be a whole discussion by themselves. That's not to say this thread is a waste though, its good to start out broadly

**Comment by roji** [comment]
@aL3891 IMHO we should not really be discussing sync vs async performance, or the ease of use of async methods - these are huge and complicated subjects. The point of this thread as far as I'm concerned is whether sync APIs which already exist in .NET Framework should be removed in corefx, and to a lesser extent (but still important), whether new classes in corefx should continue to provide both sync and async APIs.

Of course this question is related to performance and ease of use. But it's enough to acknowledge that there are some scenarios where sync I/O fits better, rather than start microbenchmarking performance and start huge philosophical discussions. This should ideally be as constrained a conversation as possible otherwise it will become unmanageable immediately.

@stephentoub, @terrajobst, @shanselman any chance you guys can weigh in on this?

**Comment by richlander** [comment]
Hi! We're in the process of revisiting some API decisions with CoreFX (more coming on that soon). There are two distinct goals:

- Make it easier for developers to adopt .NET Core.
- Enable developers writing the code they want (e.g. generic and non-generic collections, sync and async APIs). No enforced policy from 98052, just helpful guidance.

**Comment by roji** [comment]
Thanks for the info @richlander, looking forwad to hearing more details.

Somewhat ironically I'd drop non-generic collections as I can't see any advantage they offer (apart from backwards compatibility)...

**Comment by richlander** [comment]
That would be my point 1 above.

> Make it easier for developers to adopt .NET Core.

The presence of non-generic collections, for example, doesn't remove any of the value of .NET Core. For customers that want to re-use code that relies on ArrayList, they can. They are free to move that code to use List&lt;T&gt; when they feel that's a benefit to them.

**Comment by ayende** [comment]
I agree with @damageboy 
Async I/O is all well and good, but miss a _single_ ConfigureAwait(false) and suddenly you have a deadlock in your code when it is run in ASP.Net, or WPF app.

Another _major_ issue with async I/O is that they are not debuggable.
Let us assume that I've a sync system, and a request is taking a lot of time. If I'm using sync I/O, I can take a proc dump, look at the threads, and say "oh, I see that it is stuck in `ResolveDnsAddress` or some such", and I know exactly where to look.

If I'm using async I/O and tasks, I'm SOL, because there is just no way that I can track this properly.
I know that VS has some stuff that suppose to help with this, but it doesn't help for production, and it doesn't come near what we need to actually support complex systems that may exhibit issues in production.

Another issue is that if you are using slow I/O (which is a good reason to _want_ to use async), your task objects are very likely to end up in Gen1 or Gen2. That lead to more expensive GC pauses, which is something that we really care about.

I really don't like this decision, and what is more, I don't like the fact that it seemed to have been decided in darkness and just pressed onward.

**Comment by ayende** [comment]
Oh, and what is even more fun, you can't do an `await` inside a constructor, which means that in some cases, you really don't have a choice but to block.

But there is no really good way to actually do blocking that is out of the box, works across the board and doesn't incur its own set of challenges.

See also: http://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/

**Comment by JeffCyr** [comment]
@ayende 
>Another major issue with async I/O is that they are not debuggable.
Let us assume that I've a sync system, and a request is taking a lot of time. If I'm using sync I/O, I can take a proc dump, look at the threads, and say "oh, I see that it is stuck in ResolveDnsAddress or some such", and I know exactly where to look.

> If I'm using async I/O and tasks, I'm SOL, because there is just no way that I can track this properly.
I know that VS has some stuff that suppose to help with this, but it doesn't help for production, and it doesn't come near what we need to actually support complex systems that may exhibit issues in production.

It's true that async I/O is harder to debug right now. With sync code you can easily see the threads and their callstack to see what is running. With async code the data is still there, but its scattered across the heap in Task and IAsyncStateMachine objects. I've used ClrMD a couple of times to see which async methods were running in a memory dump by enumerating their associated IAsyncStateMachine. It would also be possible to reconstruct the entire async methods call tree, if Visual Studio did this to display something similar to "Parallel Stacks" it would be awesome.

>Another issue is that if you are using slow I/O (which is a good reason to want to use async), your task objects are very likely to end up in Gen1 or Gen2. That lead to more expensive GC pauses, which is something that we really care about.

This is also true, but don't you think that if you used slow I/O, a lot of other objects will end up in Gen2 (like a pending request and all associated objects) and that extra task going to Gen2 won't make a measurable impact.

**Comment by roji** [comment]
@JeffCyr,

>> Another issue is that if you are using slow I/O (which is a good reason to want to use async), your task objects are very likely to end up in Gen1 or Gen2. That lead to more expensive GC pauses, which is something that we really care about.
> This is also true, but don't you think that if you used slow I/O, a lot of other objects will end up in Gen2 (like a pending request and all associated objects) and that extra task going to Gen2 won't make a measurable impact.

That's really going to depend on your app, but in any case consider the following. If you're just thinking about that one async call, then maybe one Task on the heap seems reasonable. But any serious program is going to have a whole bunch of methods calling each other. When you're doing async, each one of those stack layers is costing a Task on the heap.

Another case that is somewhat more exotic is operations that *may* perform I/O, but in many cases won't - imagine a case where caching is implemented to save I/O operations. In these cases it's sometimes recommended to cache Task instances rather the results, to save that extra allocation. But that may not always be possible, and again, if you consider that the caching async method is deep down the stack below other async methods which don't cache - you're allocating tons of garbage *for a non-blocking operation*. Sometimes the effect can be mitigated, sometimes not, depending on what you're doing.

Remember that nobody here is saying "async is bad", just that it has its disadvantages and costs and that sync I/O should remain an option.

**Comment by GSPP** [comment]
`Task`-based async is really nasty and infectious. The two things that Go improved with over C are Garbage Collection and Green Threads for IO. It would be so nice if the CLR was 99.99% synchronous but on green threads.

It would be very rare to need to use async IO. You would only need async IO if even green threads were to expensive. That should be in the range of >100000 connected sockets or more.

Instead, we are seeing a nasty infection of "async everything" right now. Without Green Threads it's a hard tradeoff what to do. With Green Threads the decision almost always is easy: Use sync IO and block in order to not infect the caller.

**Comment by clrjunkie** [comment]
@ayende 

>Async I/O is all well and good, but miss a single ConfigureAwait(false) and suddenly you have a deadlock in your code when it is run in ASP.Net, or WPF app.

Suggest project-specific default setting for this? (e.g library project)

>Another major issue with async I/O is that they are not debuggable

Open a separate issue for this? Advocate for including APM counterpart API's ?

>Oh, and what is even more fun, you can't do an await inside a constructor, 

Advocate for a language enhancement?

What does async/await have to do with Sync Programming model vs Async Programming model..??

**Comment by ayende** [comment]
@clrjunkie Because you now asked me to open three difference issues, all of which start with -100 points.
While having those is nice, I want to know how async everything got passed that same hurdle without even seeming to think about it.

**Comment by clrjunkie** [comment]
@ayende 
>Because you now asked me to open three difference issues, all of which start with -100 points.
 
Didn't get it, is there a point system here? 

>I want to know how async everything got passed that same hurdle without even seeming to think about it.

I'll give you my take once you clarify, as I'm unclear on the "same hurdle" part.

**Comment by ayende** [comment]
@clrjunkie If all features start at -100, what made dropping sync pass that into "let's go for that".
Once you have that, just to be back in the same place, you need many other quite complex things.

**Comment by aL3891** [comment]
The points things doesn't really apply for removing stuff though, I think the way to think about it is that sync apis was a feature that also started at -100, and didn't make the cut the first time around. But again, the team are reevaluating that stuff as we speak

**Comment by ayende** [comment]
@aL3891 if you are removing stuff, you have better had a good something to replace it.
When you give async only I/O, you are losing a LOT of stuff. That need to be replaced.

Now, it made sort of sense when it was Silverlight and the browser, but when talking about "port you existing systems", that is a big bad thing to do.

**Comment by aL3891** [comment]
Well different people hold different opinions regarding what's good and what has been lost :) i'm just describing what I think their reasoning was. if enough people want it and make a good case for it, they'll add it, as it seems they are in the process of doing right now

**Comment by clrjunkie** [comment]
@ayende 
C’mon.. for all intents and purposes .NET Core is a rewrite of .NET, sure they got a head start by having the full .NET for reference and code reuse but do you really believe that the same people who wrote the code 10 – 15 years ago are same people doing the rewrite? Tell me, have you ever tried to study full .NET FW source code? Most if not all the I/O class have *two different code paths* with hoops and loops and context classes that will twist your brain, not to mention the dependency spaghetti, yes, same as you would see in any other large old codebase. 

I can totally understand why they decided to go with async/await, only for the sake of simplifying the port *for them*.  

**Implementing one code path instead of two is simpler and simplicity is an engineering criteria.**

Now the problem with this issue is the title because all what I’m reading here is people complaining about async/await short-comings thinking somehow that Sync is the solution, but that’s not necessarily true (heck I complained about it to a dead end #4834). Sync is a different *Programming Model* and I didn't read here even one scenario to justify it. 

"allocation", "ThreadPool pressure" are not Sync Api scenarios!! 

I’m not arguing Sync doesn’t have value but if it was up to me I wouldn't invest a dime in Sync unless I see a concrete use-case that shows a noticeable difference over async!

So till then, if your immediate problem is due to async/await – shout FIX ASYNC AWAIT!

**Comment by roji** [comment]
> for all intents and purposes .NET Core is a rewrite of .NET

Maybe the implementation is a rewrite, but with regards to APIs and contracts it really isn't - backwards compatibility with .NET Framework is a major (if not absolute) goal of .NET Core.

> Implementing one code path instead of two is simpler and simplicity is an engineering criteria.

@clrjunkie the cost of maintaining both a sync and an async path in any library is indeed high - you're right. I faced this exact same problem in Npgsql and ended up writing a code generator called [AsyncRewriter](https://github.com/roji/AsyncRewriter) for creating async methods from sync counterparts. There's nothing pretty about this but it does get the job done - Npgsql fully supports both sync and async I/O without me maintaining 2 full code paths.

But at the end of the day sync I/O simply isn't some small self-contained piece of functionality that can be dropped for simplifying the port for the MS team. 

Note that nobody at MS ever raised this argument - "we're dropping sync because it's too much work for us"). Nobody at MS even said "we're dropping sync", almost all sync I/O APIs are still there. It's just a sync method here and there that's missing, with a vague and uncommitted "sync is bad" argument.

> Now the problem with this issue is the title because all what I’m reading here is people complaining about async/await short-comings thinking somehow that Sync is the solution, but that’s not necessarily true (heck I complained about it to a dead end #4834). Sync is a different Programming Model and I didn't read here even one scenario to justify it.

You're looking at it the other way around. Sync isn't a "solution" to async's problems! There are simply some scenarios where async provides no advantage over sync (e.g. no scalability requirements) and only makes things worse.

Also, some of async's "problems" are inherent in the programming model - they're not something that can be fixed (your suggestions notwithstanding). And even those things that can be improved or fixed aren't there yet *now*.

If after this thread you still refuse to admit the async has its costs and that sync can *sometimes* make sense - heap allocation and GC pressure, programming model complexity, TP pressure - then you're just stuck in an ideological "async all the things" loophole that doesn't care about actual facts and programmer concerns.

**Comment by clrjunkie** [comment]
@roji
>backwards compatibility with .NET Framework is a major (if not absolute) goal of .NET Core.

Who said so? Where did you read this??

>there are simply some scenarios where async provides no advantage over sync (e.g. no scalability requirements) and only makes things worse.

No problem, since you argue so strongly about what you consider facts and programmer concerns  I challenge you to paste here 2 short console applications (async/sync) which do something useful for *you* and clearly show how sync outperforms async. That way we can discuss an actual programming use case instead of buzz words.and ideology. Ready?

**Comment by roji** [comment]
>> backwards compatibility with .NET Framework is a major (if not absolute) goal of .NET Core.

> Who said so? Where did you read this??

Read [this blog post](https://blogs.msdn.microsoft.com/dotnet/2016/02/10/porting-to-net-core/) by Immo Landwerth, and especially the section "Relationship of .NET Core and .NET Framework". .NET Core isn't a new "let's break-everything" effort.

Note also how "technologies discontinued for .NET Core" are explicitly discussed and analyzed (app domains, remoting...). Note how "sync I/O" is *not* listed in that list.

> there are simply some scenarios where async provides no advantage over sync (e.g. no scalability requirements) and only makes things worse.

> No problem, since you argue so strongly about what you consider facts and programmer concerns I challenge you to paste here 2 short console applications (async/sync) which do something useful for you and clearly show how sync outperforms async. That way we can discuss an actual programming use case instead of buzz words.and ideology. Ready?

I'm sorry, I'm done wasting my time with you on this - me (and others) have already given examples in this thread many times over. Avoiding the increased heap allocation associated with async is an *actual* performance concern for at least programmers, it's directly linked to your application's performance. Until you answer here why programmers shouldn't be concerned about heap allocations and GC pressure (and the other issues listed above) there's no point in continuing this conversation with you.

**Comment by roji** [comment]
@clrjunkie, regarding compatibility between .NET Framework and .NET Core take a look at [the corefx breaking changes guide](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md), you'll understand to what extent Microsoft is trying to make .NET Core compatible.

**Comment by ayende** [comment]
For further discussion, see this PR:
https://github.com/dotnet/corefx/pull/8642

*Hibernating Rhinos Ltd  *

Oren Eini* l CEO l *Mobile: + 972-52-548-6969

Office: +972-4-622-7811 *l *Fax: +972-153-4-622-7811

**Comment by clrjunkie** [comment]
@ayende 
For whatever reasons, this is at most an example of a problematic API design as the file handle is not polymorphic and there is a clear change in the API *functional semantics* depending how the object was constructed. Furthermore my understanding from the issue is that in this implementation Sync is simulated by waiting for Async results to complete and there is no O/S sync call involved. So the issue itself is not quite relevant to this discussion.

Saying that, I would argue that synchronous File I/O *can* be more efficient when the access pattern involves many small read/writes. I actually thought about posting here a sample program to demonstrate this, but I haven’t found a real world use-case that’s important or general enough and I didn't want to use some artificial write "counter" example.  I would bet that a work queue of fixed dedicated threads issuing many small writes *per work item* synchronously would process items faster than submitting those writes asynchronously. (assuming of course the number of work queue threads correspond to the number of CPU’s, again one work item yields many small writes). You are welcome to find *and test* a business use-case that fits this model and submit it to justify the need.

**Comment by clrjunkie** [comment]
@ayende

>Because you now asked me to open three difference issues, all of which start with -100 points, While having those is nice,

Coming to think about it, even so, if you find these truly nice then submit them, maybe they will get in someday, at least they will be documented, so do that. (No cross-reference required :)

**Comment by benaadams** [comment]
> Saying that, I would argue that synchronous File I/O can be more efficient when the access pattern involves many small read/writes. 

Nah; as you can double buffer with async and you can't with sync. So you can be preparing the next batch of data while the current one is being written out breaking the read+write dependency.

e.g. the throughput on
```csharp
private async Task CopyToAsync(Stream source, Stream destination, int bufferSize, CancellationToken cancellationToken)
{
    var buffer0 = new byte[bufferSize];
    var buffer1 = new byte[bufferSize];

    var readTask = source.ReadAsync(buffer0, 0, buffer0.Length, cancellationToken);
    var writeTask = Task.CompletedTask;

    var isOdd = false;
    int bytesRead;
    while ((bytesRead = await readTask.ConfigureAwait(false)) != 0)
    {
        readTask = source.ReadAsync(isOdd ? buffer0 : buffer1, 0, bufferSize, cancellationToken);
        isOdd = !isOdd;
        await writeTask.ConfigureAwait(false);
        writeTask = destination.WriteAsync(isOdd ? buffer0 : buffer1, 0, bytesRead, cancellationToken);
    }
    await writeTask.ConfigureAwait(false);
}
```
will be much higher than
```csharp
public void CopyTo(Stream source, Stream destination, int bufferSize)
{
    var buffer = new byte[bufferSize];
    int read;
    while ((read = source.Read(buffer, 0, buffer.Length)) != 0)
    {
        destination.Write(buffer, 0, read);
    }
}
```
Alas the framework doesn't do `CopyToAsync` this way as its more difficult to attribute failure states, e.g. what failed first.

**Comment by GSPP** [comment]
@benaadams I tried that with buffered IO. Throughput was not better because the kernel does read-aheand and write-behind in the kernel.

Also, if this *was* beneficial you could get the same benefit with sync IO and threads at a slight expense of CPU time.

**Comment by ayende** [comment]
@clrjunkie

Features implemented years from now have very little use at this point.

*Hibernating Rhinos Ltd  *

Oren Eini* l CEO l *Mobile: + 972-52-548-6969

Office: +972-4-622-7811 *l *Fax: +972-153-4-622-7811

**Comment by clrjunkie** [comment]
@benaadams 
Nah; Your biased on a particular processing model.

Each Thread:

var count= fileBatch.Count;

for(var i = 0; i < count; ++i)
{
   write(fileBatch[i]);
}

**Comment by clrjunkie** [comment]
@ayende 
>Features implemented years from now have very little use at this point.

First, you don't know what priority these feature will get. Second better late then never :)

**Comment by clrjunkie** [comment]
@benaadams 
You are assuming that items are always prepared and persisted concurrently. There could be is a time gap between the two for some use-case. I'm not ruling out any use-case.

**Comment by clrjunkie** [comment]
Buffering log records in memory into buckets - delimited, not concatenated -  and submitting them to be flushed concurrently to a set of new files after X ? Ex: A high performance logging solution that trades-off 100% durability for speed?

**Comment by roji** [comment]
@clrjunkie, regarding the question of API compatibility between .NET Core and .NET Framework, please see [this blog post from yesterday](https://github.com/dotnet/corefx/issues/5044#issuecomment-219657237). I think that pretty much closes the discussion about whether sync I/O APIs will be added. Of course you can still claim they're useless and bad.

**Comment by clrjunkie** [comment]
@roji Thanks for the heads up. (though the link you posted self references this issue..!@? )

I assume you meant this: 

https://blogs.msdn.microsoft.com/dotnet/2016/05/27/making-it-easier-to-port-to-net-core/

Well, my only "claim" is that it appears the same "rules" apply - See ya' in .NET Core V3.

Until then...Count me out on this.

-Unsubscribed-

**Comment by roji** [comment]
@clrjunkie thanks, fixed the link.

@CIPop and others at Microsoft, I'll leave this open to track bringing back SSLStream's missing sync methods, feel free to close.

**Comment by CIPop** [comment]
Thanks @roji. The scope for API surface changes that are being considered is broader than SslStream.
Regarding the async/await debugging and performance issues, I recommend opening separate issues.

Assigning this discussion to @terrajobst to track the necessary API changes.

**Comment by karelz** [comment]
@roji can you please summarize the ask? Ideally in a new issue - this one has rather long (finished) discussion hard to reason about, I would prefer to close it.

Note that the APIs will be also considered as part of .NET standard 2.0 effort (see more [here](https://blogs.msdn.microsoft.com/dotnet/2016/05/27/making-it-easier-to-port-to-net-core/) and [here](https://blogs.msdn.microsoft.com/dotnet/2016/07/15/net-core-roadmap/)).

**Comment by roji** [comment]
@karelz - the request is simply for BCL APIs to include synchronous APIs alongside asynchronous APIs, in order to support the sync programming model. A special emphasis should be made on bringing back sync APIs which exist in .NET Framework but were dropped in .NET Core 1.0 (e.g. `SSLStream.AuthenticateAsClient()`).

With the shift towards making .NET Core more API-compatible with .NET Framework to help porting, I'm not sure another issue is necessary here for that - although if you think so I'll open one.

**Comment by karelz** [comment]
Agreed that a new issue is likely not needed at this point. Thanks!
